man/xml2ssf.1
.Dd April 21, 2016
.Dt XML2SSF 1
.Os
.Sh NAME
.Nm xml2ssf
.Nd XML to BioWare SSF converter
.Sh SYNOPSIS
.Nm xml2ssf
.Op Ar options
.Op Ar input_file
.Ar output_file
.Sh DESCRIPTION
.Nm
converts XML files created by the
.Xr ssf2xml 1
tool back into the BioWare SSF format.
For a more in-depth description of SSF files,
please see the man page for the
.Xr ssf2xml 1
tool.
.Pp
The format of the input XML is pretty simple and straight-forward.
.Bd -literal
<?xml version="1.0" encoding="utf-8" standalone="yes"?>
<ssf>
  <sound id="0" label="Attack" strref="66299">vs_fkillerf_attk</sound>
  <sound id="1" label="BattleCry1">vs_fkillerf_bat1</sound>
  <sound id="2" label="BattleCry2" strref="66301"></sound>
  <sound id="3" label="BattleCry3"></sound>
</ssf>
.Ed
.Pp
The root element is
.Dq ssf ,
with multiple
.Dq sound
elements as children.
Each
.Dq sound
element requires an
.Dq id
property, specifying the index of the sound within the SSF file.
Optionally, there can be a
.Dq strref
property, which denotes the string reference (StrRef) of the text
line associated with the sound.
The contents of the
.Dq sound
element is the filename (ResRef) of the sound file to play.
.Pp
Because different BioWare games use slightly different versions of
the SSF file format, the game for which to create the SSF file has
to be specifying on the command line.
Please note that the file versions have different requirements on
the input data as well. These requirements are noted below.
.Sh OPTIONS
.Bl -tag -width xxxx -compact
.It Fl h
.It Fl Fl help
Show a help text and exit.
.It Fl Fl version
Show version information and exit.
.It Fl Fl nwn
Create an SSF file fit for use within the game
.Em Neverwinter Nights .
Sound filenames have to be 16 characters long or less.
.It Fl Fl nwn2
Create an SSF file fit for use within the game
.Em Neverwinter Nights 2 .
Sound filenames have to be 32 characters long or less.
.It Fl Fl kotor
Create an SSF file fit for use within the game
.Em Knights of the Old Republic .
No sound filenames are allowed, only StrRefs.
.It Fl Fl kotor2
Create an SSF file fit for use within the game
.Em Knights of the Old Republic II .
No sound filenames are allowed, only StrRefs.
.El
.Bl -tag -width xxxx -compact
.It Ar input_file
The XML file to convert.
If no input file is specified, the XML data is read from
.Dv stdin .
The encoding of the XML stream must always be UTF-8.
.It Ar output_file
The SSF file will be written there.
.El
.Sh EXAMPLES
Convert
.Pa file1.xml
into an SSF for Neverwinter Nights:
.Pp
.Dl $ xml2ssf --nwn file1.xml file2.ssf
.Pp
Convert
.Pa file1.xml
into an SSF for Knights of the Old Republic:
.Pp
.Dl $ xml2ssf --kotor file1.xml file2.ssf
.Sh SEE ALSO
.Xr ssf2xml 1 ,
.Xr xml2tlk 1
.Pp
More information about the xoreos project can be found on
.Lk https://xoreos.org/ "its website"
.Ns .
.Sh AUTHORS
This program is part of the xoreos-tools package, which in turn is
part of the xoreos project, and was written by the xoreos team.
Please see the
.Pa AUTHORS
file for details.


NEWS
Changes in xoreos-tools version 0.0.2
=====================================

This is the first official release of xoreos-tools, together with xoreos.

Included are:

- Tools for handling resource archives
  - unkeybif
  - unerf
  - unrim
  - unherf
  - unnds
  - desmall
- Tools converting basic resource formats into human-readable form
  - gff2xml
  - tlk2xl
  - convert2da
- Tools converting graphics formats into TGA
  - xoreostex2tga
  - unnsbtx
  - nbfs2tga
  - ncgr2tga
  - cbgt2tga
  - cdpth2tga


src/xml2ssf.cpp
/* xoreos-tools - Tools to help with xoreos development
 *
 * xoreos-tools is the legal property of its developers, whose names
 * can be found in the AUTHORS file distributed with this source
 * distribution.
 *
 * xoreos-tools is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or (at your option) any later version.
 *
 * xoreos-tools is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with xoreos-tools. If not, see <http://www.gnu.org/licenses/>.
 */

/** @file
 *  Tool to convert XML files back into SSF.
 */

#include <cstring>
#include <cstdio>

#include "src/version/version.h"

#include "src/common/scopedptr.h"
#include "src/common/ustring.h"
#include "src/common/util.h"
#include "src/common/strutil.h"
#include "src/common/error.h"
#include "src/common/platform.h"
#include "src/common/readfile.h"
#include "src/common/writefile.h"
#include "src/common/stdinstream.h"
#include "src/common/encoding.h"
#include "src/common/cli.h"

#include "src/aurora/types.h"

#include "src/xml/ssfcreator.h"

#include "src/util.h"

bool parseCommandLine(const std::vector<Common::UString> &argv, int &returnValue,
                      Common::UString &inFile, Common::UString &outFile, Aurora::GameID &game);

void createSSF(const Common::UString &inFile, const Common::UString &outFile, Aurora::GameID game);

int main(int argc, char **argv) {
	initPlatform();

	try {
		std::vector<Common::UString> args;
		Common::Platform::getParameters(argc, argv, args);

		Aurora::GameID game = Aurora::kGameIDUnknown;

		int returnValue = 1;
		Common::UString inFile, outFile;

		if (!parseCommandLine(args, returnValue, inFile, outFile, game))
			return returnValue;

		createSSF(inFile, outFile, game);
	} catch (...) {
		Common::exceptionDispatcherError();
	}

	return 0;
}

bool parseCommandLine(const std::vector<Common::UString> &argv, int &returnValue,
                      Common::UString &inFile, Common::UString &outFile, Aurora::GameID &game) {

	using Common::CLI::NoOption;
	using Common::CLI::kContinueParsing;
	using Common::CLI::Parser;
	using Common::CLI::ValGetter;
	using Common::CLI::ValAssigner;
	using Common::CLI::makeEndArgs;
	using Common::CLI::makeAssigners;
	std::vector<Common::UString> args;
	NoOption filesOpt(false, new ValGetter<std::vector<Common::UString> &>(args,
	                  "[input file] <output file>"));
	Parser parser(argv[0], "XML to BioWare SSF converter",
	              "If no input file is given, the input is read from stdin.\n\n"
	              "Since different games use different SSF file version, specifying the\n"
	              "game for which to create the SSF file is necessary.",
	              returnValue,
	              makeEndArgs(&filesOpt));

	game = Aurora::kGameIDUnknown;
	parser.addSpace();
	parser.addOption("nwn", "Create an SSF for Neverwinter Nights",
	                 kContinueParsing,
	                 makeAssigners(new ValAssigner<Aurora::GameID>(Aurora::kGameIDNWN, game)));
	parser.addOption("nwn2", "Create an SSF for Neverwinter Nights 2",
	                 kContinueParsing,
	                 makeAssigners(new ValAssigner<Aurora::GameID>(Aurora::kGameIDNWN2, game)));
	parser.addOption("kotor", "Create an SSF for Knights of the Old Republic",
	                 kContinueParsing,
	                 makeAssigners(new ValAssigner<Aurora::GameID>(Aurora::kGameIDKotOR, game)));
	parser.addOption("kotor2", "Create an SSF for Knights of the Old Republic II",
	                 kContinueParsing,
	                 makeAssigners(new ValAssigner<Aurora::GameID>(Aurora::kGameIDKotOR2, game)));


	if (!parser.process(argv))
		return false;

	if ((args.size() < 1) || (args.size() > 2) || (game == Aurora::kGameIDUnknown)) {
		parser.usage();
		returnValue = 1;

		return false;
	}

	if (args.size() == 2) {
		inFile  = args[0];
		outFile = args[1];
	} else
		outFile = args[0];

	return true;
}

void createSSF(const Common::UString &inFile, const Common::UString &outFile, Aurora::GameID game) {
	Common::WriteFile ssf(outFile);
	Common::ScopedPtr<Common::ReadStream> xml(openFileOrStdIn(inFile));

	XML::SSFCreator::create(ssf, *xml, game, inFile);

	ssf.flush();
	ssf.close();
}
